File: node/manage-custom-fields.js

const fetch = require(`node-fetch`);
const credentials = require(`./credentials.json`);
const BASE_URL = 'https://api.aweber.com/1.0/';
const client_id = credentials['clientId'];
const client_secret = credentials['clientSecret'];
const accessToken = credentials['accessToken'];

const qs = require(`querystring`);

async function request(url,options) {
    if (options == null) options = {};
    if (options.credentials == null) options.credentials = 'same-origin';
    return fetch(url, options).then(function(response) {
        if (response.status >= 200 && response.status < 300) {
            return Promise.resolve(response);
        } else {
            var error = new Error(response.statusText || response.status);
            error.response = response;
            return Promise.reject(error);
        }
    });
}

async function getCollection(accessToken, url) {
    let res;
    const collection = [];
    while (url) {
        res = await request(url, {
            headers: {
                "Authorization": `Bearer ${accessToken}`
            }
        });
        let page = await res.json();
        console.log(`got page`, {
            page
        })
        collection.push(...page.entries);
        url = page.next_collection_link;
    }
    return collection;
}

(async () => {

    // Get an account to manage custom fields on
    const accounts = await getCollection(accessToken, BASE_URL + 'accounts');
    const account = accounts[0];  // choose the first account

    // Get a list to manage custom fields on
    const listsUrl = account['lists_collection_link'];
    const lists = await getCollection(accessToken, listsUrl);
    const list = lists[0];  // choose the first list

    // Check if custom fields already exist with the names we use below
    const customFieldsUrl = list['custom_fields_collection_link'];
    const customFields = await getCollection(accessToken, customFieldsUrl);
    for (let entry of customFields) {
        if (['Test2', 'Renamed2'].includes (entry['name'])) {
            console.log (`A custom field called ${entry['name']} already exists on ${list['name']}`);
            process.exit();
        }
    }

    // Create a custom field called Test
    const createResponse = await request(customFieldsUrl, {
        method:"POST",
        'body' :qs.stringify({
            'ws.op' : 'create',
            'name' : 'Test2'
        }),
        'headers' : {
            "Access-Control-Expose-Headers": "Location",
            "Content-Type":"application/x-www-form-urlencoded",
            'Authorization' : 'Bearer ' + accessToken
        }
    });
    console.log(createResponse.headers);
    const fieldUrl = createResponse.headers.get("location");
    console.log (`Create new custom field at ${fieldUrl}`);

    // Update the custom field
    const updateResponse = await request(fieldUrl, {
        method:"patch",
        body : JSON.stringify({
            'name' : 'Renamed2',
            'is_subscriber_updateable' : true
        }),
        'headers' :{
            'Authorization' : 'Bearer ' + accessToken,
            "Content-Type":"application/json"
        }
    });
    const updatedField = await updateResponse.json();
    console.log ("Updated the custom field: ");
    console.log(updatedField);

    // Delete the custom field
    let fieldUrlResponse = await request(fieldUrl, {
        method:"delete",
        'headers' :{
            'Authorization' : 'Bearer ' + accessToken
        } 
    });
    console.log({deleteResponse:await fieldUrlResponse.json()})

    console.log ("Deleted the custom field");
})();

File: ruby/manage_custom_fields.rb
# frozen_string_literal: true

require 'faraday'
require 'json'

BASE_URL = 'https://api.aweber.com/1.0/'
credentials = JSON.parse(File.read('./credentials.json'))

conn = Faraday.new(
  headers: {
    Authorization: "Bearer #{credentials['access_token']}"
  }
) do |f|
  f.response :raise_error
end

def get_collection(conn, url)
  collection = []
  while url
    res = conn.get(url)
    page = JSON.parse(res.body)
    collection.push(*page["entries"])
    url = page["next_collection_link"]
  end
  collection
end

# Get an account to search on
accounts = get_collection(conn, "#{BASE_URL}accounts")
account = accounts.first
puts account

lists_url = account['lists_collection_link']
lists = get_collection(conn, lists_url)
list = lists.first

custom_fields_url = list['custom_fields_collection_link']
custom_fields = get_collection(conn, custom_fields_url)

custom_fields.each do |entry|
  if ["Test2","Renamed2"].include? entry['name']
    puts "A custom field called #{entry['name']} already exists on #{list['name']}"
    exit
  end
end

create_response = conn.post(
  custom_fields_url,
  'ws.op' => 'create',
  'name' => 'Test2'
) do |req|
  req.headers['Content-Type'] = 'application/x-www-form-urlencoded'
  req.headers['Access-Control-Expose-Headers'] = 'Location'
end

field_url = create_response.headers['location']
puts "Create new custom field at #{field_url}"
update_response = conn.patch(
  field_url,
  'name' => 'Renamed2',
  'is_subscriber_updateable' => true
) do |req|
  req.headers['Content-Type'] = 'application/json'
end
update_field = JSON.parse(update_response.body)
puts "Update the custom field: "
puts updated_field

field_url_response = conn.delete(field_url)
puts "delete response #{field_url_response.body}"
puts "deleted the custom field"


